rust/datafusion/src/datasource/datasource.rs

```diff
@@ -34,6 +35,23 @@ pub struct Statistics {
     pub total_byte_size: Option<usize>,
 }
 
+/// Indicates whether and how a filter expression can be handled by a
+/// TableProvider for table scans.
+#[derive(Debug, Clone)]
+pub enum TableProviderFilterPushDown {
+    /// The expression cannot be used by the provider.
+    Unsupported,
+    /// The expression can be used to help minimise the data retrieved,
+    /// but the provider cannot guarantee that all returned tuples
```

Comment:
I think those are good names and good docs as well

Reply:
This result is used to drive the logical plan optimisation. Returning `Unsupported` _or_ `Inexact` will both result in the original filter node being preserved in the plan (only `Exact` can provide optimisation guarantees that the post-hoc filtering is unecessary) but `Inexact` means that the filter expr in question will be passed to `TableProvider::scan` when creating the execution plan, whereas `Unsupported` means the provider can do nothing useful with the expr at all (relevant code is [here](https://github.com/apache/arrow/pull/8917/files#diff-5f6ab9edcef3a4b9a8eb75f641ac0930cdbae92156f056fea400e0c58218a094R354-R369)).

Reply:
I think that these names make sense as does the behavior described

Reply:
I agree the names explicit! 

I am challenging whether having 3 values is necessary. `Unsupported` vs `Inexact` is something that you don't need to expose publicly here as they are used internally by the the TableProvider when they will be passed back to the `scan()` method (or not). Said otherwise, you can leave it to the the implementation of `scan()` to decide what to do with the expressions, you don't need do do it with `supports_filter_pushdown()`. Only `Exact` vs `Unsupported|Inexact` really needs to be exposed because it is used externally (by the planner)

Reply:
Right, I see what you mean now :) Yes, from a semantics point-of-view, that would result in the same results being delivered. The major issue is that this would add filters to scan nodes unnecessarily and there's currently quite a lot of exact string matching for plan outputs in tests, so this would break lots of existing code.

Reply:
I see... that's annoying! how many tests fail approximately? 

If tests influence our design, we should be sure that they are doing so for a reason (not just because they are poorly written). Here I feel that maybe we should fix them, no? In #8910, I already tweaked an API because otherwise I would have had lots of tests to change (more details [here](https://github.com/apache/arrow/pull/8910#discussion_r543331995)). This was a mistake (due mainly to laziness 🙃), but here I believe the interface is too important to do the same 😄. Any opinion @alamb ?

Reply:
Basically I don't have a strong opinion either way -- putting in three enum variants when one may be of limited utility seems ok to me -- we could always remove / collapse them later

@rdettai  -- what do you think about merging this PR in and then proposing a clean up  in a subsequent PR? I feel like the design by @returnString is pretty good, even if one of the enum values is of limited utility.

Reply:
I leave it to you to decide, it won't take me long to adapt my code to these changes anyway.